### PR TITLE
OpenPaaS-Suite/esn#10 Build SPA for production using webpack + babel

### PR DESCRIPTION
**Pull request by chamerling**

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -5,15 +5,22 @@
   "main": "index.js",
   "scripts": {
     "test": "echo \"Error: no test specified\" && exit 1",
-    "serve": "webpack-dev-server --watch --progress",
-    "build:dev": "webpack"
+    "serve": "webpack-dev-server --watch --progress --env=dev",
+    "serve:build": "webpack-dev-server --env=prod",
```
Comment (MichaelBailly) [inline package.json:9]: rename to "serve:prod"

File: package.json [hunk 2]
```diff
@@ -5,15 +5,22 @@
...
+    "build": "webpack --env=prod"
```
Comment (MichaelBailly) [inline package.json:10]: bring back "build:dev" plz

**Comment by huy-ta** [comment]
Can you check if this Webpack's configuration works with `npm link` (_e.g._ linking from `esn-frontend-calendar` to the local `esn-frontend-common-libs`)? Or is there any alternative if one wants to test some local changes?

**Comment by chamerling** [comment]
> Can you check if this Webpack's dev configuration works with `npm link` (_e.g._ linking from `esn-frontend-calendar` to the local `esn-frontend-common-libs`)? Or is there any alternative if one wants to test some local changes?

@huy-ta dev is still working as it is working now. There are no problem with npm links currently. The only issue is for build in prod mode where is can fail, but we can say that we will not build prod stuff with npm links

**Comment by chamerling** [comment]
> Look good!
> 
> Just a question: Why did not to have `babel` in `dev` mode?
> 
> > Babel is used to convert ECMAScript 2015+ code into a backwards compatible version of JavaScript in current and older browsers or environments
> 
> It means, dev mode still needs it, no?

@tuanlc for now babel is mainly used for ng-annotate in prod mode. We should not have any issue in dev mode until we need some unreleased ECMAScript feature.

**Comment by tuanlc** [comment]
> @tuanlc for now babel is mainly used for ng-annotate in prod mode. We should not have any issue in dev mode until we need some unreleased ECMAScript feature.

As I saw, we have used some ES6 syntax like `const`. Not sure where we use `unreleased ECMAScript`, but it really cool if dev can use advantages from them like we did for server side.

**Comment by chamerling** [comment]
> As I saw, we have used some ES6 syntax like `const`. Not sure where we use `unreleased ECMAScript`, but it really cool if dev can use advantages from them like we did for server side.

Allright, current browser you probably use for develop is probably already supporting everything we need (https://kangax.github.io/compat-table/es6/). By not adding babel to dev, we avoid to transpile for nothing, and winning some seconds each time we update files. Anyway, if you think we really need it, we can add it in dev.

File: webpack.commons.js [hunk 3]
```diff
@@ -0,0 +1,249 @@
...
+      },
+    ],
+  },
+  devServer: {
+    contentBase: path.join(__dirname, 'dist'),
+    contentBasePublicPath: BASE_HREF,
+    compress: true,
+    port: 9900,
+    proxy: [{
+      context: [
```
Comment (piavgh) [inline webpack.commons.js:222]: We should update this array. Each SPA will have a different list of URLs.
But I think we can do this in Polish week?

**Comment by huy-ta** [comment]
> > As I saw, we have used some ES6 syntax like `const`. Not sure where we use `unreleased ECMAScript`, but it really cool if dev can use advantages from them like we did for server side.
> 
> Allright, current browser you probably use for develop is probably already supporting everything we need (https://kangax.github.io/compat-table/es6/). By not adding babel to dev, we avoid to transpile for nothing, and winning some seconds each time we update files. Anyway, if you think we really need it, we can add it in dev.

I don't think we need to use Babel for development, unless you really want to use bleeding-edge ECMAScript features. But I doubt anyone will need that.

File: webpack.commons.js (hunk 3 above)
Comment (chamerling) [reply webpack.commons.js:222]: yes for polish

**Pull request merged**